src/gadgets/ecc_circuit.rs

```diff
@@ -0,0 +1,540 @@
+#![allow(non_snake_case)]
+use crate::gadgets::utils::{alloc_num_equals, alloc_one, alloc_zero, conditionally_select};
+use bellperson::{
+  gadgets::{
+    boolean::{AllocatedBit, Boolean},
+    num::AllocatedNum,
+    Assignment,
+  },
+  ConstraintSystem, SynthesisError,
+};
+use ff::PrimeField;
+use rand::rngs::OsRng;
+
+#[derive(Clone)]
+pub struct AllocatedPoint<Fp>
+where
+  Fp: PrimeField,
+{
+  x: AllocatedNum<Fp>,
+  y: AllocatedNum<Fp>,
+  is_infinity: AllocatedNum<Fp>, //TODO: Make this Allocatedbit?
```

Comment:
Yes, make it an `AllocatedBit`. That will make boolean operations much cheaper (and much less error-prone).